### PR TITLE
EVP: Let EVP_PKEY_gen() initialize ctx->keygen_info

### LINKED ISSUE #12047: dsaparam and dhparm are broken
Trying to repair the tests in `test/recipes/90-test_store.t` for #11733 
I've come across nice crashes of `dsaparam`.

Calling it with a key size acceptable these days, e.g., 2048:
```
apps/openssl dsaparam  -out dsaparam.pem 2048
```
or
```
apps/openssl dsaparam -genkey -out dsakey.pem 2048
```
one gets for instance
```
==11521==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7f6239ee36d8 bp 0x7fff02e120f0 sp 0x7fff02e12050 T0)
==11521==The signal is caused by a WRITE memory access.
==11521==Hint: address points to the zero page.
    #0 0x7f6239ee36d8 in ossl_callback_to_pkey_gencb crypto/evp/pmeth_gn.c:136
    #1 0x7f623a127516 in dsa_gencb providers/implementations/keymgmt/dsa_kmgmt.c:487
    #2 0x7f6239d002f5 in BN_GENCB_call crypto/bn/bn_prime.c:115
    #3 0x7f6239ef4091 in generate_q_fips186_4 crypto/ffc/ffc_params_generate.c:290
    #4 0x7f6239ef54e5 in ffc_params_FIPS186_4_gen_verify crypto/ffc/ffc_params_generate.c:634
    #5 0x7f6239ef71b3 in ffc_params_FIPS186_4_generate crypto/ffc/ffc_params_generate.c:982
    #6 0x7f6239db6b1f in dsa_generate_ffc_parameters crypto/dsa/dsa_gen.c:46
    #7 0x7f623a1279b5 in dsa_gen providers/implementations/keymgmt/dsa_kmgmt.c:528
    #8 0x7f6239ec5b58 in evp_keymgmt_gen crypto/evp/keymgmt_meth.c:335
    #9 0x7f6239ec3587 in evp_keymgmt_util_gen crypto/evp/keymgmt_lib.c:413
    #10 0x7f6239ee3cf8 in EVP_PKEY_gen crypto/evp/pmeth_gn.c:190
    #11 0x7f6239ee4358 in EVP_PKEY_paramgen crypto/evp/pmeth_gn.c:276
    #12 0x44daa7 in dsaparam_main apps/dsaparam.c:184
    #13 0x465fc1 in do_cmd apps/openssl.c:467
    #14 0x46528c in main apps/openssl.c:279
```
It turns out that `ossl_callback_to_pkey_gencb()` tries to assign
```
    ctx->keygen_info[0] = p;
    ctx->keygen_info[1] = n;
```
while `ctx->keygen_info == NULL` because the `ctx` is not properly initialized.
I tried to fix this myself but this goes beyond my OpenSSL knowledge.
It looks like this is due to a mistake in the recent switch to the provider mechanism,
and unfortunately there is no test at least at CLI level that would have caught this.

### PR DESCRIPTION
**Pull request by levitte**
In EVP_PKEY_METHOD code, the backend initializes ctx->keygen_info.
With provider side code, it's not possible to reach back into the
EVP_PKEY_CTX in the same manner, so we need to make that
initialization in the central generation function, EVP_PKEY_gen().

This isn't quite compatible with the idea that keygen_info could have
an arbitrary amount of elements, but since all our legacy backends use
exactly two elements, that's what we go for.

Fixes #12047

## APPS: Fix 'openssl dsaparam -genkey'

Using a parameter EVP_PKEY for key generation with EVP_PKEY routines
works a little differently than the raw DSA routines that were used
before.

While fixing that, clean away all remaining use of the DSA type, which
simplifies the code a bit more.

## APPS: Fix 'openssl dhparam'

'dhparam' can't be completely rewritten in terms of EVP_PKEY functions
yet, because we lack X9.42 support.  However, we do when generating,
but forgot to extract a DH pointer with EVP_PKEY_get0_DH().

### CODE REVIEW

File: crypto/evp/pmeth_gn.c [hunk 1]
```diff
@@ -165,6 +167,17 @@ int EVP_PKEY_gen(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey)
...
+    /*
+     * Asssigning gentmp to ctx->keygen_info is something our legacy
+     * implementations do.  Because the provider implementations aren't
+     * allowed to reach into our EVP_PKEY_CTX, we need to provide similar
+     * space for backward compatibility.  It's ok that we attach a local
+     * variable, as it should only be useful in the calls down from here.
+     * This is cleared as soon as it isn't useful any more, i.e. directly
+     * after the evp_keymgmt_util_gen() call.
+     */
+    ctx->keygen_info = gentmp;
```
Comment (DDvO) [inline crypto/evp/pmeth_gn.c:179]: Maybe also add `ctx->keygen_info_count = 2` ?

Comment (levitte) [reply crypto/evp/pmeth_gn.c:179]: Ah, right, forgot about that!

**Comment by levitte** [comment]
I didn't try out the `-genkey`, done now, and discovered that the conversion of `dsaparam` to use EVP_PKEY was...  lacking a bit.  Passing the parameter EVP_PKEY to the EVP_PKEY key generation works *differently* than with raw DSA routines...  Fix commit coming up!

File: crypto/evp/pmeth_gn.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/pmeth_gn.c:179]: Done

**Comment by levitte** [comment]
`-genkey` issue fixed as well

**Comment by DDvO** [comment]
> I didn't try out the `-genkey`, done now, and discovered that the conversion of `dsaparam` to use EVP_PKEY was... lacking a bit. Passing the parameter EVP_PKEY to the EVP_PKEY key generation works _differently_ than with raw DSA routines... Fix commit coming up!

Now `dsaparam` crashes again, at least for me, this time for any key size I try.
```
apps/openssl dsaparam -out param.pem 1024

AddressSanitizer:DEADLYSIGNAL
=================================================================
==26286==ERROR: AddressSanitizer: SEGV on unknown address 0x000008eff532 (pc 0x7f56b90d2bc8 bp 0x7ffc21e07260 sp 0x7ffc21e07140 T0)
==26286==The signal is caused by a READ memory access.
    #0 0x7f56b90d2bc8 in EVP_PKEY_get_utf8_string_param crypto/evp/p_lib.c:1862
    #1 0x7f56b90ddf14 in EVP_PKEY_gen crypto/evp/pmeth_gn.c:231
    #2 0x7f56b90de44e in EVP_PKEY_paramgen crypto/evp/pmeth_gn.c:292
    #3 0x44da9b in dsaparam_main apps/dsaparam.c:183
    #4 0x465ea7 in do_cmd apps/openssl.c:467
    #5 0x465172 in main apps/openssl.c:279
    #6 0x7f56b8ae809a in __libc_start_main ../csu/libc-start.c:308
```

File: apps/dsaparam.c [hunk 2]
```diff
@@ -66,9 +66,8 @@ const OPTIONS dsaparam_options[] = {
 int dsaparam_main(int argc, char **argv)
 {
     ENGINE *e = NULL;
-    DSA *dsa = NULL;
     BIO *in = NULL, *out = NULL;
-    EVP_PKEY *pkey = NULL;
+    EVP_PKEY *params, *pkey = NULL;
```
Comment (DDvO) [inline apps/dsaparam.c:70]: gcc in a Travis run correctly pointed out that `params` needs to be initializied (to `NULL`).
Then the crash disappears, and everything seems to work fine :)

**Comment by DDvO** [comment]
gcc in a Travis run correctly points out that `params` needs to be initializied.
When I init it to `NULL` the crash disappears, and everything seems to work fine :)

File: apps/dsaparam.c (hunk 2 above)
Comment (DDvO) [inline apps/dsaparam.c:70]: gcc in a Travis run correctly pointed out that `params` needs to be initialized (to `NULL`).
Then the crash disappears, and everything seems to work fine :)

File: apps/dsaparam.c [hunk 3]
```diff
@@ -181,51 +180,34 @@ int dsaparam_main(int argc, char **argv)
                        "Error, DSA key generation setting bit length failed\n");
             goto end;
         }
-        if (EVP_PKEY_paramgen(ctx, &pkey) <= 0) {
+        if (EVP_PKEY_paramgen(ctx, &params) <= 0) {
```
Comment (DDvO) [inline apps/dsaparam.c:183]: nice cleanup of the variable name.

**Comment by levitte** [comment]
> Please also provide to the user a to-the-point error message in case the key size is unsuitable.

That will happen separately.  There are a number of places where we've been sloppy, and we should make a concise sweep.

File: apps/dsaparam.c (hunk 2 above)
Comment (levitte) [reply apps/dsaparam.c]: Fixed

**Comment by levitte** [comment]
The only Travis failure was a timeout, assumed unrelated.
(we have now migrated to new Travis, the one used for this PR can't be restarted)

**Comment by levitte** [comment]
Merged

4ec1463d71db6324abe9c91d2ed9aa1e136c9cb3 EVP: Let EVP_PKEY_gen() initialize ctx->keygen_info
edf96591650551274c650a48c228d089c4057df9 APPS: Fix 'openssl dsaparam -genkey'
78215852066d214ded6695a27c997eb0d651c31f APPS: Fix 'openssl dhparam'

**Pull request closed without merging**